rfd/0197-better-async-state-management-with-tanstack-query.md

```diff
@@ -0,0 +1,266 @@
+loading and error states - they're handled by the component tree.
+
+There are other benefits to using Suspense - it gives more control over the loading state when there are multiple
+components that need to fetch data, as well as allowing for more granular error handling. [Read more here](https://react.dev/reference/react/Suspense).
+
+### Example query
+
+```tsx
+const { data, error, isPending } = useQuery({
+  queryKey: ['accessList', accessListId],
+  queryFn: () => accessManagementService.fetchAccessList(accessListId),
+});
+```
+
+Which could be wrapped as a hook for easy reuse:
+
+```tsx
+export function useGetAccessList(id: string) {
+  return useQuery({
+    queryKey: ['accessList', id],
+    queryFn: () => accessManagementService.fetchAccessList(id),
+  });
+}
+```
+
+### Example mutation
+
+```tsx
+const { mutateAsync: deleteAccessList, isPending } = useMutation({
+  mutationFn: (accessListId: string) =>
+    accessManagementService.deleteAccessList(accessListId),
+});
+
+async function onOk() {
+  await deleteAccessList(accessListId);
+
+  history.push(cfg.getAccessListManagementRoute(), {
+    deletedAccessListId: accessListId,
+  });
+}
+```
+
+### Example infinite query
+
+```tsx
+const { fetchNextPage, hasNextPage, data, isPending } = useInfiniteQuery({
+  getNextPageParam: lastPage => lastPage.next_cursor,
+  initialPageParam: '',
+  queryFn,
+  queryKey,
+});
+```
+
+This allows the next page to be fetched through `fetchNextPage` without having to worry about the pagination cursor,
+as well as an indication as to whether there is more data to fetch.
+
+### Example paginated query
+
+TanStack Query allows us to easily create paginated queries (and even pre-fetch the next page of results for faster
+rendering),
+as well as keeping the previous page's data whilst the next page is being fetched. This is not easily
+achievable with the current `useAttempt` or `useAsync` methods.
+
+```tsx
+const [page, setPage] = useState(0);
+
+const { data, error, isFetching, isPlaceholderData } = useQuery({
+  queryKey: ['users', page],
+  queryFn: () => fetchUsers(page),
+  placeholderData: keepPreviousData,
+})
+
+useEffect(() => {
+  if (!isPlaceholderData && data?.next_cursor) {
+    queryClient.prefetchQuery({
+      queryKey: ['users', page + 1],
+      queryFn: () => fetchUsers(page + 1),
+    })
+  }
+}, [data, isPlaceholderData, page, queryClient])
+```
+
+## Example Implementation
+
+https://github.com/gravitational/teleport/pull/47282 shows an example of how we can introduce TanStack Query into the
+codebase.
+
+This can be incremental, with new pages using TanStack Query and old pages being converted over time.
+
+Moving away from passing the result from `useAttempt` (or the "attempt" returned from `useAsync`) results in better
+tested
+code (the network requests will have to be mocked, which reflects the real world usage), as well as better stories
+reflecting the actual state of the page,
```

Comment:
i believe a lot of the newer tests just mock the backend response with msw or something instead of passing around fake attempts so hopefully this pattern will just be used going forward. its better to mock the api response anyway rather than ovewriting whatever fetching mechanism is being used so I like this path more